src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    const uint8_t* address;
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int result = gai_error(&state->gai_request);
+    HostEntry* entry = state->entry;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+
+    if (result == 0)
+    {
+        const uint8_t* address = state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, entry);
+    }
+
+    free(state);
+
+    if (callback != NULL)
+    {
+        callback(entry, ret);
+    }
+}
+#endif
+
+static void GetHostEntryForNameCreateHint(struct addrinfo* hint)
+{
+    // Get all address families and the canonical name
+
+    memset(hint, 0, sizeof(struct addrinfo));
+    hint->ai_family = AF_UNSPEC;
+    hint->ai_flags = AI_CANONNAME;
+}
+
+int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
+{
+#if HAVE_GETADDRINFO_A
+    return 1;
+#else
+    return 0;
+#endif
+}
+
+int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entry)
+{
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo hint;
+    GetHostEntryForNameCreateHint(&hint);
+
+    struct addrinfo* info = NULL;
+
+    int result = getaddrinfo((const char*)address, NULL, &hint, &info);
+    if (result != 0)
+    {
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return GetHostEntries(address, info, entry);
+}
+
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, HostEntry* entry, GetHostEntryForNameCallback callback)
+{
+#if HAVE_GETADDRINFO_A
+    if (address == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo hint;
+    GetHostEntryForNameCreateHint(&hint);
```

Comment:
This `hint` structure will not last the length of the async operation. I'm not sure if it is required to.

Reply:
Hm, then the native tests (added temporarily with this PR) should fail too?!

Anyway, I'll change this to be on the safe side and the struct lasts the whole operation (will pack it into `GetAddrInfoAsyncState`).

Maybe that's the obvious failure I didn't see. Thanks therefore.

Reply:
Unfortunately no change when `hint` isn't on the stack but a member of the heap-allocation struct `GetAddrInfoAsyncState`. Still crashes.

Reply:
This is const, so could be a global static constant. (See other comment.)